chore: RFC for emitting multiple log events from remap (#7038)
timberio/vector: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).

Additionally, I think we can leverage the approach I describe here in the future to allow users to emit different data types than the incoming type. For example, users could emit logs related to an incoming metric, or metrics from an incoming log.

## Internal Proposal

This is roughly the same as https://github.com/timberio/vector/issues/6330#issuecomment-772562955 with some slight tweaks.

A new `emit_log` function will be added to the VRL stdlib:

```text
emit_log(value: Object)
```

This function will cause the object passed as value to be emitted at that point and flushed downstream. The emitted log will have its metadata copied from the input event.

> Additionally, an `emit_root` (we can work on the naming) config option will be added to the `remap` transform to configure whether `.` is emitted after the transform runs. It will default to `true` to preserve the current behavior but can be set to `false` by users to supress this behavior. Admittedly, I'm not wild about introducing this additional config option, but I'm not seeing another great alternative.

Reviewer: Could this be a flag set at runtime in the script?

This will be able to be combined with the iteration mechanism that will be introduced [#6031](https://github.com/timberio/vector/issues/6031) to emit an unknown number of events. Naively this might look something like:

```text
for stooge in .stooges

>    emit(stooge)

Reviewer: ITYM `emit_log`

end
```

In the future we can also add functions for emitting metrics like:

```text
emit_counter(namespace: String, name: String, timestamp: Timestamp, value: Float, kind: "absolute"|"relative")
```

> I considered having just an `emit_metric()` but it would require users to pass in objects that match exactly the internal representation we have for metrics.

Reviewer: Not necessarily. The Lua wrapper has at various points converted to/from the internal format as necessary, as does the WASM wrapper. There's no reason this couldn't either.

Author: Yeah, that's true, but we don't have a concept of structs in VRL so it would depend on the user passing a map of key/values that matches some schema.

## Doc-level Proposal

The `remap` tranform would gain an extra configuration option:

```text
emit_root = true/false # default false
```

When `emit_root` is `true`, the value of `.` will be emitted at the end of the remap program. When `emit_root` is false, the value of `.` will not be emitted. Instead users should use the `emit_log` function to emit.

## Rationale

This enhances remap to support additional transformation use-cases of splitting up an event into multiple events. Without this, users will continue to have to use Lua to acheive this, which introduces a performance bottleneck compared to remap

## Prior Art

- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)

These are similar to the `explode` transform PR we have: https://github.com/timberio/vector/pull/6545

## Drawbacks

- It makes the remap transform more complicated to understand and adds an additional configuration option that users need to be aware of
- Ongoing maintenance burden should be minimal

## Alternatives

### Not emitting . if an `emit_log` is present in the program

> We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.

Reviewer: I was going to suggest this, but I agree it is a potentially surprising behavior. It also may introduce variable behavior if the script does call `emit_log` in some code path.

This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.

This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.

I also don't see an easy way to extend this to deal with emitting multiple metrics without requiring users to create objects with a known format (like `{ "name": "foo", "namespace": "bar", "type = "counter", "value": 1 ...}`) whereas it seems more straightforward to extend the proposed approach with functions like `emit_counter`

### An `explode` transformation

https://github.com/timberio/vector/pull/6545

Have a separate, special purpose, transform for converting a single log event into multiple log events.

This would work, but I think the downsides are:

- Introduces a new transform rather than just leveraging an existing one, remap. There is some precedence for this with the `route` and `reduce` transforms though.
- It is less flexible than letting users emit arbitrary events

> - There isn't a clear path to emitting, say, multiple metrics from an input log event

Reviewer 2: If we _don't_ want to introduce a new transform, then I'd probably lean towards updating the `remap` transform with a new option `explode = true`. If set, the transform would check at boot that your VRL script returns an array, and it would behave the same as what I described above, except that it doesn't require a separate transform and allows you to mutate the event before you explode it into zero or more events.

I still lean towards a separate transform being the best option, but this would probably be my second preference.

I realize that both of my proposals are _less flexible_ than an `emit_log`, but I think overall they come out on top due to their simplicity over the current proposal in the RFC.

> - There isn't a clear path to emitting, say, multiple metrics from an input log event

Author: Thanks for the alternate proposal!

I like the idea of a separate transform where it evaluates the last expression (similar to conditions) to get the output. This should avoid the need to decouple `.` from the underlying event for now.

> I still somewhat prefer a separate transform for this. I worry that — in a couple of months — we'll be at a point where the remap transform is a "do anything you want" transform, which it was never meant to be.

Agreed, I can see that this is expanding the scope significantly. I think I do see a future where remap is more general purpose, similar to Lua, to offer increased flexibility, but we can defer that for now until we have more data from users about what they are trying to achieve.

> > The biggest open question for me with this approach is whether the event in this transform should be _immutable_, to enforce people to use `remap` to mutate events, and `explode` to convert to multiple events, instead of using `explode` to do both.
>
> I'd probably lean to **yes**, make the event immutable, if nothing else, it gives us more flexibility down the line on what we want to do with this transform, and it also guides operators to use our transforms "for their intended purpose". We can always remove that restriction in the future, if it turns out to be a good idea to do so.

I also lean towards yes, it should be immutable in that modifications should not persist. I do think it should be possible to do some mapping in the `explode` transform though. For example, I think it will be common to want to map over the elements that are being exploded to add additional fields like:

Input:

```
{ "host": "foobar", "events": [{ "message": "foo" }, { "message": "bar" }] }
```

```toml
[transforms.explode]
type = "explode"
source = "map(array!(.events), |event| event.host = .host) ?? []"
```

The exact "mapping over elements" syntax will be contingent on #6031 .

I'll update the RFC to reflect this approach as I think it is a simpler start and we can always add in `emit_log` later if we find an eed.

{"message": "foo"}
{"message": "bar"}
```

If any elements in the returned array are not an object, they will be turned into a string and set as the `message` key on event.

## Rationale

This enhances Vector with a native transform capable of transforming an event into multiple events. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to this proposal.

## Prior Art

- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)

These are similar to the proposed approach.

## Drawbacks

- Adds an additional transform to be aware of.
- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.
- This idea of the last line of the remap script being the "return value" could be a bit confusing to users in how it differs from `remap` which uses whatever `.` is at the end of the script. However, this is similar to VRL's use in unit tests and conditions for the `filter` and `route` transforms.
- Ongoing maintenance burden should be minimal.

> ## Alternatives

Reviewer 3: One other alternative could be a config option on the remap transform like `unnest_field = "events"` that works vaguely like SQL's `UNNEST` operator and is applied after the script is run. 

We built remap to reduce the amount that users had to chain together primitive transforms to do what they want, so this might be a little bit more in that spirit than adding a new primitive transform.

   emit_log(stooge)
end
```

In the future we can also add functions for emitting metrics like:

```text
emit_counter(namespace: String, name: String, timestamp: Timestamp, value: Float, kind: "absolute"|"relative")
```

I considered having just an `emit_metric()` but it would require users to pass in objects that match exactly the internal representation we have for metrics.

The `remap` tranform would gain an extra configuration option:

```text
emit_root = true/false # default false
```

When `emit_root` is `true`, the value of `.` will be emitted at the end of the remap program. When `emit_root` is false, the value of `.` will not be emitted. Instead users should use the `emit_log` function to emit.

We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.

### Modifying remap to accept setting the root object to an array

https://github.com/timberio/vector/issues/6988

This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.

> This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.

Reviewer 3: Would this still be difficult if Vector's `Event` could be an array at the top level?

> # RFC 6330 - 2021-04-07 - Emitting multiple log events from remap transform

Reviewer 2: ```suggestion
# RFC 6330 - 2021-04-07 - Emitting multiple log events from explode transform
```

> This RFC describes an approach for emitting multiple log events from a remap transform.

Reviewer 2: ```suggestion
This RFC describes an approach for emitting multiple log events from an `explore` transform.
```

> [Credit to @JeanMertz](https://github.com/timberio/vector/pull/7038#discussion_r609476064) for this revised approach.
> 

Reviewer 2: ```suggestion
```

If I had to give credits to everyone that helped me over the years, you'd be reading a novel for every PR or RFC I'd write 😄

## Scope

In:

- Turning a log event into multiple log events

Out:

- Turning a log event into multiple metrics
- Turning a metric into multiple metrics
- Turning a metric into multiple logs

## Motivation

> Currently the remap transform can only modify an event as it is passing through. Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.

Reviewer 2: ```suggestion
Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
```

I don't think it's relevant to mention the remap transform in this case, as it is complimentary, but serves a different purpose.

Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).

## Internal Proposal

> The proposal is to add an `explode` transform that makes use of remap to emit a set of events from one input event by requiring the remap script to resolve to an array. For each element of the array, a separate event will be published.

Reviewer 2: ```suggestion
The proposal is to add an `explode` transform that makes use of Vectir Remap Language (VRL) to emit a set of events from one input event by requiring the VRL program to resolve to an array. For each element of the array, a separate event will be published.
```

Example input:

```json
{ "events": [{ "message": "foo" }, { "message": "bar" }] }
```

Transform config:

```toml
[transforms.explode]
type = "explode"
source = "array!(.events) ?? []" # will be typechecked at compile-time
```

Ouput:

```json
{"message": "foo"}
{"message": "bar"}
```

Support for iteration as part of [#6031](https://github.com/timberio/vector/issues/6031) will allow for users to do things like map fields onto each element. An example might look something like:

Input:

```json
{ "host": "foobar", "events": [{ "message": "foo" }, { "message": "bar" }] }
```

Transform config (actual mapping syntax TBA):

```toml
[transforms.explode]
type = "explode"
source = "map(array!(.events), |event| event.host = .host) ?? []"
```

Output:

```json
{"host": "foobar", "message": "foo"}
{"host": "foobar", "message": "bar"}

> ```

Reviewer 2: Is this relevant to mention in this RFC? It seems orthogonal.

Also, it might actually be _more performant_ to do this in a separate `remap` transform, which can then/eventually run in parallel for multiple events.

Author: True, but how would they get the fields from the original source event in the later remap transform? I think they'd need to do the mapping in the `explode` transform.

This is similar to the support that the current [`explode` transform PR](https://github.com/timberio/vector/pull/6545) has for merging in top-level fields when creating events from a subfield that has an array.

## Doc-level Proposal

> To convert an incoming event into multiple events, the `explode` transform can be used. This transform takes a VRL script where it expects the last expression to return an array.

Reviewer 2: Just syncing terminology used throughout our docs/code.

```suggestion
To convert an incoming event into multiple events, the `explode` transform can be used. This transform takes a VRL program where it expects the last expression to return an array.
```

For example, given an input of:

```json
{ "events": [{ "message": "foo" }, { "message": "bar" }] }
```

And a transform of:

```toml
[transforms.explode]
type = "explode"
source = "array!(.events) ?? []"
```

The following events will be output:

```json
{"message": "foo"}
{"message": "bar"}
```

## Rationale

> This enhances vector with a native transform capable of an event into multiple events. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to remap

Reviewer: ```suggestion
This enhances vector with a native transform capable of an event into multiple events. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to remap.
```

> This enhances vector with a native transform capable of an event into multiple events. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to remap

Reviewer 2: ```suggestion
This enhances Vector with a native transform capable of transforming an event into multiple events. Without this, users will continue to have to use Lua or WASM to achieve this, which introduces a performance bottleneck compared to a native transform.
```

## Prior Art

- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)

These are similar to the proposed approach.

## Drawbacks

- Adds an additional transform to be aware of.
- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.

> - This idea of the last line of the remap script being the "return value" could be a bit confusing to users in how it differs from `remap` which uses whatever `.` is at the end of the script.

Reviewer 2: This is true, but it's worth pointing out that it _is_ similar to how VRL-based conditions work (in unit-tests, but also filter and route transforms).

We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.

### Modifying remap to accept setting the root object to an array

https://github.com/timberio/vector/issues/6988

This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.

This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.

Using a separate `explode` transform keeps the responsibilities of the transform more clear and avoids having to refactor the `remap` transform to decouple `.` from the underlying Vector `Event` object; though we may still want to do this in the future anyway.

## Open Questions

> 1. How should we handle the case where the user returns an array where one or more elements is not an object? My instinct for now is just to log and drop them.

Reviewer: I can't think of anything else reasonable to do other than dropping the whole batch, and that seems excessively harsh.

> 1. How should we handle the case where the user returns an array where one or more elements is not an object? My instinct for now is just to log and drop them.

Reviewer 2: My proposal would be to handle types as follows:

- Objects are turned into individual events.
- Any other type is turned into an individual event, with the value of the type set as the `message`.
- We can make this configurable through a transform config if we so desire.

E.g.

```json
{ "events": [{ "timestamp": "...", "message": "foo" }, "bar", 42.0] }
```

Turns into:

```jsonl
{ "message": "foo", "timestamp": "..." }
{ "message": "bar" }
{ "message": 42.0 }
```

If operators want, they can either use a filter transform or a remap transform to shape events to their liking after this happens (e.g. enforce message becomes a string in the case of the last one, or filter out events without a timestamp in the case of the second and third one).

Author: 👍 I like this.

# RFC 6330 - 2021-04-07 - Converting one log event into multiple log events

This RFC describes an approach for transforming one input log event into multiple log events.

## Scope

In:

- Turning a log event into multiple log events

Out:

- Turning a log event into multiple metrics
- Turning a metric into multiple metrics
- Turning a metric into multiple logs

## Motivation

Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.

Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).

## Internal Proposal

The proposal is to:

1. Extend `remap` so that if `.` is an array at the end, it will emit one event for each element in that array
2. Add an `unnest` VRL function that will transform an object into an array of objects using a specified field on the input object

> 3. Add an `unnest` transform that can be used for simple cases (could just "expand" into a `remap` transform)

Reviewer 4: I would vote against a separate transform just for consistency and simplicity when browsing transforms.

Author: I tend to agree. I added it just based on earlier feedback on this proposal, but I'm inclined to leave it out. We can always add it later if the need arises.

Reviewer 2: We can leave it off for now. I do agree if we go with the `.` solution the added transform is less of a requirement.

If we went with the previous proposal of using the final value of the program to determine the events triggered, then we'd need a new transform, since that's how how `remap` currently works.

Example input:

```json
{ "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }] }
```

Remap transform config:

```toml
[transforms.remap]
type = "remap"
source = """
. = unnest(., "events")
"""
```

Ouput:

```json
{ "host": "localhost", "message": "foo" }
{ "host": "localhost", "message": "bar" }
```

This would also be available as an `unnest` transform for simple cases where the user does not need to do any additional mapping:

```toml
[transforms.unnest]
type = "unnest"
field = "events"
```

Which would have the same effect. I think the implementation could just take advantage of the "expand" feature of transforms that lets them replace themselves at vector start time with another transform.

> Additionally, we will provide `only_fields` and `except_fields` as options on the `unnest` function and transform to allow users to select which fields will be kept. These match similar semantics to the `encoding` options on sinks.

Reviewer 2: Is there a big need for this, compared to using `del` before calling `unnest`? I guess these arguments could be considered an easier solution to achieve the same, but I'm not convinced it's worth the added complexity.

Author: Yeah, it's not a big deal, users could workaround this via calling `del` beforehand. It's just a shortcut.

Though, we don't have an easy way to say "delete all fields _but_ these" yet.

Example input:

```json
{ "timestamp": "2020-12-09T16:09:53+00:00", "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }] }
```

Remap transform config:

```toml
[transforms.remap]
type = "remap"
source = """

> . = unnest(., "events", only_fields: ["host"])

Reviewer 2: This is the first time (I believe) where we use a string to represent a path. Is there a reason why you chose this over a path-type?, e.g.:

```suggestion
. = unnest(., .events, only_fields: [.host])
```

This would bring it more in-line with functions like `del`, which also takes a path, but doesn't expand that path to a value, instead using it to manipulate the target event.

There are edge-cases to consider here (e.g. what happens when you provide an indexed-path), but overall it does seem to me like the better choice.

Author: Ah, yes, I think I like that. I had actually forgotten we could just take the path without resolving it.

We could even just do:

```
. = unnest(.events)
```

Where it would just pull `events` up one level since the path sufficient information about the parent.

This does reduce the user's ability to dynamically set the parameter based on some other value though (e.g. using a different value for `only_fields` by setting a variable prior).

I could see users being slightly confused that:

```
foo = [.host]
. = unnest(., .events, only_fields = foo)
```

Doesn't work. `delete` has this same behavior though.

I'll think about this a bit more.

It is now occurring to me that `del` is also a bit special since it modifies the argument by reference. I think that is the only function that does that.

Author: We can discuss this more in the PR for `unnest`, but I'm inclined to just stick with strings rather than requiring paths. I think this'll be less surprising to users.

Reviewer 2: > It is now occurring to me that `del` is also a bit special since it modifies the argument by reference. I think that is the only function that does that.

That is correct. This was discussed during implementation. There were a few more that directly modified the target (such as `only_fields`), but that one was removed before the 1.0 release.

Ergonomically, it felt more useful to have `del` behave this way, and return the value of the deleted field, as it made renaming fields easier to do (`.foo = del(.bar)`).

> We can discuss this more in the PR for `unnest`, but I'm inclined to just stick with strings rather than requiring paths. I think this'll be less surprising to users.

I agree, we don't need to decide here, but I don't think we should use paths, or at least if we do we should be consistent and make this part of the specification (https://github.com/timberio/vector/issues/3740).

If any elements in the array field are not an object, they will be turned into a string and set as the `message` key.

The `remap` transform can also be used to emit multiple events from a single incoming event by setting the root path, `.`, to an array. The above transform is synonymous with:

```toml
[transforms.remap]
type = "remap"
source = """
. = unnest(., "events")
"""
```

## Rationale

This enhances `remap` to be able to emit multiple events as well as provides a `unnest` transform for simple cases. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to this proposal.

## Prior Art

- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)

These are similar to the proposed approach.

## Drawbacks

- Adds an additional transform to be aware of.
- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.

> - Ongoing maintenance burden should be minimal.

Reviewer 2: An additional drawback of this solution as opposed to the previous one (separate transform, using final value of program), is that it _can_ result in unexpected outcomes if someone is doing something like this:

```jsonl
// most common events
{ "message": "{ \"foo\": true }" }

// uncommon, unexpected
{ "message": "[true, false]" }
```

```swift
. = parse_json!(.message)
```

output:

```jsonl
// expected
{ "foo": true }
```

```jsonl
// unexpected
{ "message": true }
{ "message": false }
```

By having exploding/unnesting be done in a dedicated transform, people explicitly opt-in to this behaviour, which can still cause unexpected results for badly formatted events, but it means (1) that when you see multiple events you know exactly where to debug, and (2) when using the `remap` transform, you never have to worry about the above situation occurring.

This is why I'm still in favour of the previous proposal from a useability perspective, but I'm not _opposed_ to the current approach, so I'm happy to see this land if others prefer this over the alternative.

Reviewer 2: I should note that the above example is something I've seen happen plenty of times as I worked on similar observability systems in the past (e.g. tens/hundreds of services, managed by different teams, using their own techniques/libraries to emit events, some untyped, and accidentally emitting non-conforming event shapes).

Author: 🤔 That is a good observation.

I think generally people neglect that `parse_json` can return other things than just objects. I think maybe we could improve the docs for that function as a start to have examples of enforcing the expected type through the type assertion functions.

I agree a separate transform would make it more obvious, but I also like the idea of users just being able to use remap.

Reviewer 5: One approach to address this could be adding an explicit flag (e.g. `emit_multiple`) to the transform config, which would error when anything other than an array of objects is present in the VRL root after transformation.

Or vice versa error when an array is present but the flag was not set.

We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.

### Modifying remap to accept setting the root object to an array

https://github.com/timberio/vector/issues/6988

This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.

This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.

Using a separate `explode` transform keeps the responsibilities of the transform more clear and avoids having to refactor the `remap` transform to decouple `.` from the underlying Vector `Event` object; though we may still want to do this in the future anyway.

## Open Questions

> 1. Do we want the additional `unnest` transform? I included it based on [feedback that using VRL was offputting for simple cases](https://github.com/timberio/vector/pull/7038#issuecomment-817043491), but maybe the newly proposed `remap` modifications are easy enough to use that we don't need the additional transform.

Reviewer 4: Ha, the current Ben disagrees with the past Ben. My stance has changed. Assigning `.` to an array has always struck me as the simplest approach.
